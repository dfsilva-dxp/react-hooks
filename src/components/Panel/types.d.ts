import { ReactNode } from "react";

export interface IPanelProps {
  title?: string;
  children?: ReactNode;
}
