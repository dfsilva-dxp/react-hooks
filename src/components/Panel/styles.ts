import styled, { css } from "styled-components";

export const PanelWrapper = styled.div`
  ${({ theme }) => css`
    width: 100%;
    height: calc(100vh - 4rem);
    background-color: ${theme.colors.elements};
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xsmall};
  `}
`;

export const PanelHeader = styled.header`
  ${({ theme }) => css`
    padding-bottom: ${theme.spacings.xsmall};
    border-bottom: 1px solid ${theme.colors.background};
  `}
`;

export const PanelBody = styled.div`
  ${({ theme }) => css`
    padding-top: 0;

    header + & {
      padding-top: ${theme.spacings.xsmall};
    }
  `}
`;
