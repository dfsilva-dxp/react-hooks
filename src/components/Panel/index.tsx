import * as Styled from "./styles";

import { IPanelProps } from "./types";

export default function Panel({ children, title }: IPanelProps) {
  return (
    <Styled.PanelWrapper>
      {!!title && (
        <Styled.PanelHeading>
          <h4>{title}</h4>
        </Styled.PanelHeading>
      )}

      <Styled.PanelBody>{children}</Styled.PanelBody>
    </Styled.PanelWrapper>
  );
}
