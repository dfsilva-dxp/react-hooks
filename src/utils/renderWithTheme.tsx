import { ReactNode } from "react";
import { render, RenderResult } from "@testing-library/react";
import { ThemeProvider } from "styled-components";

import { MyTheme } from "../styles";

export const renderWithTheme = (children: ReactNode): RenderResult =>
  render(<ThemeProvider theme={MyTheme}>{children}</ThemeProvider>);
