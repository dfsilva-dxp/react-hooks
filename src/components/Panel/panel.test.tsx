import { screen } from "@testing-library/react";

import { renderWithTheme } from "../../utils/tests/renderWithTheme";

import Panel from ".";

describe("<Panel/>", () => {
  it("should render currectly", () => {
    renderWithTheme(
      <Panel>
        <span>Item</span>
      </Panel>
    );

    const body = screen.getByText(/item/i).parentElement;
    const wrap = body?.parentElement;

    expect(wrap).toHaveStyle({
      width: "100%",
      height: "calc(100vh - 4rem)",
      "background-color": "#202024",
      "border-radius": "0.25rem"
    });

    expect(body).toHaveStyle({
      "padding-top": "0",
      display: "grid",
      "grid-template-columns": "1fr"
    });

    expect(body).toHaveStyleRule("grid-template-columns", "repeat(2,1fr)", {
      media: "(min-width: 768px)"
    });
    expect(body).toHaveStyleRule("grid-template-columns", "repeat(3,1fr)", {
      media: "(min-width: 1200px)"
    });
    expect(body).toHaveStyleRule("grid-template-columns", "repeat(4,1fr)", {
      media: "(min-width: 1400px)"
    });
  });

  it("should render title when passed", () => {
    const { container } = renderWithTheme(
      <Panel title="React Hooks">
        <span>Item</span>
      </Panel>
    );

    const heading = screen.getByRole("heading", { name: /react hooks/i });
    const body = screen.getByText(/item/i).parentElement;

    expect(heading).toBeInTheDocument();
    expect(heading.parentElement).toHaveStyle({
      "padding-bottom": "1rem",
      "border-bottom": "1px solid #121214"
    });

    expect(body).toHaveStyle({
      "padding-top": "1rem"
    });

    expect(container).toMatchSnapshot();
  });
});
