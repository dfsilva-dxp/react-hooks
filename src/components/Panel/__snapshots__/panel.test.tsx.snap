// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Panel/> should render title when passed 1`] = `
.c0 {
  width: 100%;
  height: calc(100vh - 4rem);
  background-color: #202024;
  border-radius: 0.25rem;
  padding: 1rem;
}

.c1 {
  padding-bottom: 1rem;
  border-bottom: 1px solid #121214;
}

.c3 {
  padding-top: 0;
  display: grid;
  grid-template-columns: 1fr;
  gap: 1rem;
}

header + .c2 {
  padding-top: 1rem;
}

@media (min-width:768px) {
  .c3 {
    grid-template-columns: repeat(2,1fr);
  }
}

@media (min-width:1200px) {
  .c3 {
    grid-template-columns: repeat(3,1fr);
  }
}

@media (min-width:1400px) {
  .c3 {
    grid-template-columns: repeat(4,1fr);
  }
}

<div>
  <div
    class="c0"
  >
    <header
      class="c1"
    >
      <h4>
        React Hooks
      </h4>
    </header>
    <div
      class="c2 c3"
    >
      <span>
        Item
      </span>
    </div>
  </div>
</div>
`;
